OffscreenCanvas does not work as TexImageSource in WebGL

kkinnunen-apple opened the pull request:
#### 4ca89ddeb5a2a28a9cfdb1e3bdd2891c79930b51
<pre>
OffscreenCanvas does not work as TexImageSource in WebGL
<a href="https://bugs.webkit.org/show_bug.cgi?id=253886">https://bugs.webkit.org/show_bug.cgi?id=253886</a>
rdar://106700463

Reviewed by Matt Woodrow.

Add texImageSource overload for OffscreenCanvas. Later on the Context2D,
WebGL and OffscreenCanvas codepaths could be unified and made faster.

* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGL2RenderingContext.idl:
* Source/WebCore/html/canvas/WebGLRenderingContext.idl:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSource):
(WebCore::WebGLRenderingContextBase::validateOffscreenCanvas):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* LayoutTests/TestExpectations:
Explicitly add pass on tests that would timeout with OffscreenCanvas
not working correctly.

Canonical link: <a href="https://commits.webkit.org/262009@main">https://commits.webkit.org/262009@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d4ded0a0ad62d1e14abe1c25ef9e52157c0bdb1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/99 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/101 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/81 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/94 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/93 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/75 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/164 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/84 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/77 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/91 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/86 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/75 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/72 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/60 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/90 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -3793,6 +3792,32 @@ ExceptionOr<void> WebGLRenderingContextBase::texImageSource(TexImageFunctionID f
 #endif
 
+#if ENABLE(OFFSCREEN_CANVAS)
+ExceptionOr<void> WebGLRenderingContextBase::texImageSource(TexImageFunctionID functionID, GCGLenum target, GCGLint level, GCGLint internalformat, GCGLint border, GCGLenum format, GCGLenum type, GCGLint xoffset, GCGLint yoffset, GCGLint zoffset, const IntRect& inputSourceImageRect, GCGLsizei depth, GCGLint unpackImageHeight, OffscreenCanvas& source)
```
Is there any reason we can't have a single codepath for `CanvasBase` instead of separate ones for `HTMLCanvasElement` and `OffscreenCanvas`?

kkinnunen-apple replied:
Yeah, there's no common method to access the pixels.
There should be, but that could be for the next patch.

mattwoodrow replied:
`buffer` and `copiedImage` are both on `CanvasBase`, what else would we need?

kkinnunen-apple committed:
OffscreenCanvas does not work as TexImageSource in WebGL
https://bugs.webkit.org/show_bug.cgi?id=253886
rdar://106700463

Reviewed by Matt Woodrow.

Add texImageSource overload for OffscreenCanvas. Later on the Context2D,
WebGL and OffscreenCanvas codepaths could be unified and made faster.

* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGL2RenderingContext.idl:
* Source/WebCore/html/canvas/WebGLRenderingContext.idl:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSource):
(WebCore::WebGLRenderingContextBase::validateOffscreenCanvas):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* LayoutTests/TestExpectations:
Explicitly add pass on tests that would timeout with OffscreenCanvas
not working correctly.

Canonical link: https://commits.webkit.org/262009@main
